homeassistant/components/sensor/recorder.py

```diff
@@ -146,6 +147,48 @@ def _time_weighted_average(
     return accumulated / period_seconds
 
 
+def _time_weighted_circular_mean(
+    fstates: list[tuple[float, State]], start: datetime.datetime, end: datetime.datetime
+) -> float:
+    """Calculate a time weighted circular mean.
+
+    The circular mean is calculated by weighting the states by duration in seconds between
+    state changes.
+    Note: there's no interpolation of values between state changes.
+    """
+    old_fstate: float | None = None
+    old_start_time: datetime.datetime | None = None
+    values: list[float] = []
+
+    for fstate, state in fstates:
+        # The recorder will give us the last known state, which may be well
+        # before the requested start time for the statistics
+        start_time = max(state.last_updated, start)
+        if old_start_time is None:
+            # Adjust start time, if there was no last known state
+            start = start_time
+        else:
+            duration = start_time - old_start_time
+            # Append same value for each second between state changes
+            assert old_fstate is not None
+            values.extend(old_fstate for _ in range(int(duration.total_seconds())))
```

Comment:
This is not necessary, and also not correct because we truncate the durations.

Calculating a circular mean is essentially adding vectors of length 1, and then calculating the angle as explained here https://en.wikipedia.org/wiki/Circular_mean#Definition

So, we can just do a scalar multiplication of the angles with the duration instead, the helper function could be updated like this to facilitate that:
```py
def circular_mean(values: list[float, float]) -> float:
    """Return the circular mean of the values."""
    sin_sum = sum(math.sin(math.radians(x)) * duration for x, duration in values)
    cos_sum = sum(math.cos(math.radians(x)) * duration for x, duration in values)
    return math.degrees(math.atan2(sin_sum, cos_sum)) % 360
```